src/libponyrt/actor/actor.c

```diff
@@ -19,31 +19,148 @@
+}
+
+static void set_internal_flag(pony_actor_t* actor, uint8_t flag)
+{
+  actor->internal_flags = actor->internal_flags | flag;
+}
+
+static void unset_internal_flag(pony_actor_t* actor, uint8_t flag)
+{
+  actor->internal_flags = actor->internal_flags & (uint8_t)~flag;
+}
+
+//
+// Mute/Unmute/Check mute status functions
+//
+// For backpressure related muting and unmuting to work correctly, the following
+// rules have to be maintained.
+//
+// 1. Across schedulers, an actor should never been seen as muted when it is not
+// in fact muted.
+// 2. It's ok for a muted actor to be seen as unmuted in a transient fashion
+// across actors
+//
+// If rule #1 is violated, we might end up deadlocking because an actor was
+// muted for sending to an actor that might never be unmuted (because it isn't
+// muted). The actor muted actor would continue to remain muted and the actor
+// incorrectly seen as muted became actually muted and then unmuted.
+//
+// If rule #2 is violated, then a muted actor will receive from 1 to a few
+// additional messages and the sender won't be muted. As this is a transient
+// situtation that should be shortly rectified, there's no harm done.
+//
+// Our handling of atomic operations in `mute_actor`
+// and `unmute_actor` are to assure that both rules aren't violated.
+
+static void mute_actor(pony_actor_t* actor)
+{
+  set_sync_flag(actor, SYNC_FLAG_MUTED);
+  DTRACE1(ACTOR_MUTED, (uintptr_t)actor);
+}
+
+void ponyint_unmute_actor(pony_actor_t* actor)
+{
+  unset_sync_flag(actor, SYNC_FLAG_MUTED);
+  DTRACE1(ACTOR_UNMUTED, (uintptr_t)actor);
+}
 
-bool has_flag(pony_actor_t* actor, uint8_t flag)
+static bool triggers_muting(pony_actor_t* actor)
 {
-  uint8_t flags = atomic_load_explicit(&actor->flags, memory_order_relaxed);
-  return (flags & flag) != 0;
+  return has_sync_flag_any(actor, SYNC_FLAG_OVERLOADED |
+    SYNC_FLAG_UNDER_PRESSURE | SYNC_FLAG_MUTED);
 }
 
-static void set_flag(pony_actor_t* actor, uint8_t flag)
+static void actor_setoverloaded(pony_actor_t* actor)
 {
-  uint8_t flags = atomic_load_explicit(&actor->flags, memory_order_relaxed);
-  atomic_store_explicit(&actor->flags, flags | flag, memory_order_relaxed);
+  pony_assert(!ponyint_is_cycle(actor));
+  set_sync_flag(actor, SYNC_FLAG_OVERLOADED);
+  DTRACE1(ACTOR_OVERLOADED, (uintptr_t)actor);
 }
 
-static void unset_flag(pony_actor_t* actor, uint8_t flag)
+static void actor_unsetoverloaded(pony_actor_t* actor)
 {
-  uint8_t flags = atomic_load_explicit(&actor->flags, memory_order_relaxed);
-  atomic_store_explicit(&actor->flags, flags & (uint8_t)~flag,
-    memory_order_relaxed);
+  pony_ctx_t* ctx = pony_ctx();
+  unset_sync_flag(actor, SYNC_FLAG_OVERLOADED);
+  DTRACE1(ACTOR_OVERLOADED_CLEARED, (uintptr_t)actor);
+  if (!has_sync_flag(actor, SYNC_FLAG_UNDER_PRESSURE))
+  {
+    ponyint_sched_start_global_unmute(ctx->scheduler->index, actor);
+  }
+}
+
+static void maybe_mute(pony_ctx_t* ctx, pony_actor_t* to)
```

Comment:
this appears to be the logic for ponyint_maybe mute?

Reply:
note that ponyint_maybe_mute and maybe_mute where two different functions and you appear to have turned them all into ponyint_maybe_mute calls while "renaming it" to `maybe_mute` which isn't correct.

Reply:
`maybe_mute` was only called from one place and i removed/replaced the code and renamed `ponyint_maybe_mute` to `maybe_mute` as part of https://github.com/ponylang/ponyc/pull/4135/commits/121f0311830d5436489018f17fe72b1e82b0ee7c (one of those last two commits are code cleanup and performance related that can be removed/split into their own PRs if desired that i mentioned in my earlier comment).

Seems like my two cleanup/performance related commits are causing confusion in reviewing this PR. i think i'll revert them and create separate PRs once this once has been settled.